design/volume-group-snapshot.md

```diff
@@ -0,0 +1,504 @@
+  - This label would act as a group identifier for the PVCs to be grouped under a specific VGS.
+  - It will be used to collect the PVCs to be used for a particular instance of VGS object.  
+**Note:** Modifying or adding VGS label on PVCs during an active backup operation may lead to unexpected or undesirable backup results. To avoid inconsistencies, ensure PVC labels remain unchanged throughout the backup execution.
+
+#### Changes to the Existing PVC ItemBlockAction plugin:
+  - Currently the PVC IBA plugin is applied to PVCs and adds the RelatedItems for the particular PVC into the ItemBlock.
+  - At first it checks whether the PVC is bound and VolumeName is non-empty.
+  - Then it adds the related PV under the list of relatedItems.
+  - Following on, the plugin adds the pods mounting the PVC as relatedItems.
+  - Now we need to extend this PVC IBA plugin to add the PVCs to be grouped for a particular VGS object, so that they are processed together under an ItemBlock by Velero.
+      - First we will check if the PVC that is being processed by the plugin has the user specified VGS label.
+      - If it is present then we will execute a List call in the namespace with the label as a matching criteria and see if this results in any PVCs (other than the current one).
+      - If there are PVCs matching the criteria then we add the PVCs to the relatedItems list.
+      - This helps in building the ItemBlock we need for VGS processing, i.e. we have the relevant pods and PVCs in the ItemBlock.
+
+**Note:** The ItemBlock to VGS relationship will not always be 1:1. There might be scenarios when the ItemBlock might have multiple VGS instances associated with it.
+Lets go over some ItemBlock/VGS scenarios that we might encounter and visualize them for clarity:
+1. Pod Mounts: Pod1 mounts both PVC1 and PVC2.  
+   Grouping: PVC1 and PVC2 share the same group label (group: A)  
+   ItemBlock: The item block includes Pod1, PVC1, and PVC2.  
+   VolumeGroupSnapshot (VGS): Because PVC1 and PVC2 are grouped together by their label, they trigger the creation of a single VGS (labeled with group: A).  
+
+```mermaid
+flowchart TD
+   subgraph ItemBlock
+   P1[Pod1]
+   PVC1[PVC1 group: A]
+   PVC2[PVC2 group: A]
+   end
+
+   P1 -->|mounts| PVC1
+   P1 -->|mounts| PVC2
+
+   PVC1 --- PVC2
+
+   PVC1 -- "group: A" --> VGS[VGS group: A]
+   PVC2 -- "group: A" --> VGS
+
+```   
+2. Pod Mounts: Pod1 mounts each of the four PVCs.  
+   Grouping:  
+     Group A: PVC1 and PVC2 share the same grouping label (group: A).  
+     Group B: PVC3 and PVC4 share the grouping label (group: B)   
+   ItemBlock: All objects (Pod1, PVC1, PVC2, PVC3, and PVC4) are collected into a single item block.   
+   VolumeGroupSnapshots:  
+      PVC1 and PVC2 (group A) point to the same VGS (VGS (group: A)).      
+      PVC3 and PVC4 (group B) point to a different VGS (VGS (group: B)).    
+```mermaid
+flowchart TD
+    subgraph ItemBlock
+        P1[Pod1]
+        PVC1[PVC1 group: A]
```

Comment:
one nuance I could not find in the doc. 
Grouping can only be done at driver type. 
Say PVC1,2,3 are all in Group A
But PVC 1,2 are driver D1 and 3 is driver D2

We can't group these 3 PVCs under one VGS.
As per my understanding each driver needs to impl the VGS API.
Or is VGS agnostic of driver and will call into different drivers for different snapshots.

This has further design implication IMO.

Let me know if this is already covered somewhere and I missed it.

Reply:
@anshulahuja98 yes I think you are correct, you cannot mix and match PVCs for a particular VGS instance. It depends on the VGSClass specified in the VGS object spec. 
Shall I just add a note regarding this caveat in the design ?

Reply:
I think this might require a design enhancement instead of just a note. 

you can take this 2 ways, 1 way is to assume customer grouped the volumes properly per each driver and if you encounter different drivers you fail the VGS and fallback to normal 1 snapshot at a time. 

2nd way is during VGS creation create different VGS for each driver. For Group A, create 2 VGS. 

Open to other approaches. 

Sideline - 
We should also add a "note" for scenarios where all drivers don't support VGS.

Reply:
I think velero should create VGS per driver, and label it such that more than one VGS can be listed by velero for a PVC group for restore.

Reply:
@kaovilai so if 4 PVCs had the same label, 2 for driver1 and 2 for driver2, we'd have two different VGS created? I wonder how that works on the VGS side -- if you have 2 VGS created with the same label selector. I'm guessing they'd each have a different VolumeGroupSnapshotClass? And does the VGS controller automatically filter on driver when creating the volume snapshots?

Reply:
> I'm guessing they'd each have a different VolumeGroupSnapshotClass?

yes.

> And does the VGS controller automatically filter on driver when creating the volume snapshots?

does k8s not have mechanism to pick the right VGSC per driver already? but yes we can lean on prior volumesnapshotclass work where we had label there per driver.

Reply:
@kaovilai The filter on driver question was more about the PVCs. i.e. if I have PVCs ceph1 and aws1 and both have the "group=foo" label, if I create a VGS with a ceph VGSClass and "group=foo" -- will the VGS controller filter the PVC list to skip aws1 since it's not a ceph PVC?

I guess we don't need to know that impl detail in Velero, though -- in this case, Velero will wait for 2 VS to be created. If we only get 1 by the CSISnapshotTimeout, then we bail. If we get 2, then the VS with mismatched VGSC/VSC will itself fail.

Reply:
@anshulahuja98 So I think the missing part is that the VGS isn't really driver-agnostic, since there's an associated VolumeGroupSnapshotClass which should identify the driver. If the user erroneously labels a volume which is incompatible with the VGSClass, then I'm guessing that there won't be a VS created for that volume. If this is the case, then the plugin will time out waiting for the VS to be created, resulting in a partially failed backup. 
@shubham-pampattiwar is my understanding correct here? We probably need to be more explicit about this in the design. If it's not 100% clear what the CSI snapshotter will do if the group includes incompatible PVCs, we may need to leave some open questions in the  design that will be resolved in future testing.

Reply:
We have control over the spec of the VGSClass to be used in VGS spec. I am thinking of the following steps:
- Identify the CSI driver from the group of PVCs 
- If there are PVCs with more than once CSI driver we skip VGS creation and fail the backup if the PVC and Velero backup goes to PartiallyFailed state
- If there is only one CSI driver for all the PVCs, continue with VGS creation and use the VGSClass where the CSI driver matches with the grouped PVC CSI driver.
WDYGT ?

Reply:
@shubham-pampattiwar so in this case, when we identify that the PVC has a group label, we list the PVCs with the same label -- and if they don't all have the same driver, then we error out on this PVC immediately? (same process will repeat on the other PVCs in the group) -- so then if we have a VG with 2 PVCs, backup will be PartiallyFailed with both PVCs erroring out with a descriptive message saying that you can't group PVCs with different provisioners?

Reply:
Looking in my cluster now, I see that the PV has `spec.csi.driver` and the PVC has `volume.kubernetes.io/storage-provisioner` -- both have the same value in this instance. Can we just use the annotation here, so we don't have to get the PV for every PVC here?

Reply:
@sseago yup I think we can use that, and fallback on inspecting PV if that annotation is missing.
cc: @anshulahuja98 @kaovilai @Lyndon-Li @blackpiglet

Reply:
Updated the design, PTAL, Thanks !

Reply:
Instead of erroring and going to ParitallyFailed, can we doing fallback to Single CSI snapshot for that PVC?

Reply:
@anshulahuja98 I think the problem is that the user has identified several volumes that must be snapshotted as a group for consistency reasons. If Velero cannot do that, we should explicitly error out rather than silently failing to do what was asked. User can then react to the failure by updating the labels, changing CSI drivers, etc. depending on how important it is for the volumes to be snapshotted together. We do need to make sure that the error message is clear enough to allow the user to fix it -- i.e. list the PVCs in the group and indicate that they do not all share the same provisioner.